### PR TITLE
feat: publish storage metrics to billing provider

### PR DESCRIPTION
**Pull request by adilansari**
## Describe your changes
- Publish storage metrics to Metronome
- Removes checked in mocks, they are now generated on demand

## How best to test these changes
- Unit tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -99,7 +104,28 @@ coverdir:
...
+generate_mocks:
+	for ifaces in $(MOCK_INTERFACES) ; do \
+		MOCK_DIR=$${ifaces%:*} ; \
+		MOCK_NAME="$${ifaces#*:}" ; \
+		mockery --dir $${MOCK_DIR} --name $${MOCK_NAME} ; \
+  	done
+
+# clean mocks
+clean_mocks:
+	for ifaces in $(MOCK_INTERFACES) ; do \
```
Comment (efirs) [inline Makefile:122]: That's why I am saying that having all the mock in one top level package would be easier.
Clean would be just rm -f mock/*
In generate MOCK_DIR also would be static.

Comment (adilansari) [reply Makefile:122]: I am running into circular imports that way if interfaceB uses interfaceA:

```
mocks/
    mock_interfaceB.go   // imports packageA/packageB
packageA/
    test_interfaceA.go.  // imports packageB && mocks
    packageB/
         interfaceB.go
         others.go
```

File: server/metadata/tenant.go [hunk 2]
```diff
@@ -374,6 +375,19 @@ func (m *TenantManager) GetTenant(ctx context.Context, namespaceId string) (*Ten
 	return tenant, nil
 }
 
+func (m *TenantManager) AllTenants(_ context.Context) []*Tenant {
+	m.RLock()
+	defer m.RUnlock()
+
+	tenants := make([]*Tenant, len(m.tenants))
+	i := 0
+	for _, t := range m.tenants {
```
Comment (himank) [inline server/metadata/tenant.go:384]: just do `for i, t := range m.tenants {}`

File: server/services/v1/billing/event.go [hunk 3]
```diff
@@ -17,6 +17,8 @@ package billing
 import (
 	"time"
 
+	"strconv"
```
Comment (himank) [inline server/services/v1/billing/event.go:20]: goimports

File: server/services/v1/billing/event.go [hunk 4]
```diff
@@ -132,10 +144,20 @@ func (sb *StorageEventBuilder) Build() *StorageEvent {
...
-	billingMetric.Timestamp = sb.timestamp
+	// default time
+	if sb.timestamp.IsZero() {
+		sb.timestamp = time.Now()
+	}
+	billingMetric.Timestamp = sb.timestamp.Format(TimeFormat)
 	props := make(map[string]interface{})
 
+	// auto generate transaction id
+	if len(sb.transactionId) != 0 {
```
Comment (himank) [inline server/services/v1/billing/event.go:155]: Is this during retries?

File: server/services/v1/billing/reporter.go [hunk 5]
```diff
@@ -138,3 +116,71 @@ func (r *UsageReporter) push() error {
...
+			WithNamespaceId(nsMeta.StrId).
+			WithTimestamp(time.Now()).
+			WithDatabaseBytes(dbBytes).
+			Build()
+		events = append(events, event)
+
+		log.Debug().Msgf("reporting storage for %s -> %+v", nsMeta.StrId, event)
+	}
+
+	err := r.billingSvc.PushStorageEvents(r.ctx, events)
```
Comment (himank) [inline server/services/v1/billing/reporter.go:157]: `return r.billingSvc.PushStorageEvents(r.ctx, events)`

File: server/metadata/tenant.go (hunk 2 above)
Comment (adilansari) [reply server/metadata/tenant.go:384]: map

File: server/services/v1/billing/event.go (hunk 4 above)
Comment (adilansari) [reply server/services/v1/billing/event.go:155]: Just to make sure transaction id is unique at a particular time. Duplicate will be ignored anyways.

Comment (himank) [reply server/services/v1/billing/event.go:154]: I was wondering when it is coming from `sb` vs when we need to generate it.

**Review by himank — APPROVED** [review]

**Pull request merged**